src/peft/tuners/lora/layer.py

```diff
@@ -149,6 +152,41 @@ def reset_lora_parameters(self, adapter_name, init_lora_weights):
             nn.init.zeros_(self.lora_embedding_A[adapter_name])
             nn.init.normal_(self.lora_embedding_B[adapter_name])
 
+    def pissa_init(self, adapter_name, init_lora_weights):
+        weight = self.get_base_layer().weight
+        dtype = weight.dtype
+        if dtype not in [torch.float32, torch.float16, torch.bfloat16]:
+            raise TypeError(
+                "Please initialize PISSA at full precision before quantizing the residual model, which may reduce quantization errors. For more details, see https://github.com/fxmeng/peft/blob/47f8534701ae4c164aa50af0f6e31dc8f37f7857/examples/pissa_finetuning/finetune_pissa_in_4bit.py."
+            )
+        weight = weight.to(torch.float32)
+
+        if init_lora_weights == "pissa":
+            # USV^T = W <-> VSU^T = W^T, where W^T = weight.data in R^{out_channel, in_channel},
+            V, S, Uh = torch.linalg.svd(weight.data, full_matrices=False)
+            Vr = V[:, : self.r[adapter_name]]
+            Sr = S[: self.r[adapter_name]]
+            Sr /= self.scaling[adapter_name]
+            Uhr = Uh[: self.r[adapter_name]]
+        elif len(init_lora_weights.split("_niter_")) == 2:
+            Vr, Sr, Ur = svd_lowrank(
+                weight.data, self.r[adapter_name], niter=int(init_lora_weights.split("_niter_")[-1])
+            )
+            Sr /= self.scaling[adapter_name]
+            Uhr = Ur.t()
+        else:
+            raise ValueError(
+                f"init_lora_weights should be 'pissa' or 'pissa_niter_[number of iters]', got {init_lora_weights} instead."
+            )
+
+        lora_A = torch.diag(torch.sqrt(Sr)) @ Uhr
+        lora_B = Vr @ torch.diag(torch.sqrt(Sr))
+        self.lora_A[adapter_name].weight.data = lora_A
+        self.lora_B[adapter_name].weight.data = lora_B
+        weight = weight.data - self.scaling[adapter_name] * lora_B @ lora_A
+        weight = weight.to(dtype)
+        self.get_base_layer().weight.data = weight
```

Comment:
Out of curiosity, did you run experiments where the base weights were not updated?

Reply:
> updated

Thank you for your suggestion. Unlike LoRA, which does not modify the residual model, directly incorporating the principal singular values and vectors will significantly alter the function of the model, potentially compromising its capabilities. However, if we adjust the scaling of the principal singular values and vectors to a relatively small value, positive outcomes may emerge. We will verify this possibility in future ablation studies.

Reply:
> Out of curiosity, did you run experiments where the base weights were not updated?

I conducted an experiment using the PiSSA adapter in conjunction with the base model for training. Due to PiSSA altering the functionality of the base model, the initial steps showed very high training loss and grad_norm, with initial values around 0.7 and 7, respectively, in the PiSSA+residual model. However, after 100 iterations, the loss surprisingly converged to around 5. This training will be completed by tomorrow, and I look forward to seeing the final outcomes then.

1-10 steps:
```
{'loss': 13.4825, 'grad_norm': 11509.810546875, 'learning_rate': 8.333333333333333e-07, 'epoch': 0.0}                                                                                                                                                                         
{'loss': 13.4192, 'grad_norm': 16469.279296875, 'learning_rate': 1.6666666666666667e-06, 'epoch': 0.0}                                                                                                                                                                        
{'loss': 13.457, 'grad_norm': 15556.3232421875, 'learning_rate': 2.5e-06, 'epoch': 0.0}                                                                                                                                                                                       
{'loss': 13.3041, 'grad_norm': 3012.66943359375, 'learning_rate': 3.3333333333333333e-06, 'epoch': 0.01}                                                                                                                                                                      
{'loss': 12.9197, 'grad_norm': 1206.873291015625, 'learning_rate': 4.166666666666667e-06, 'epoch': 0.01}                                                                                                                                                                      
{'loss': 12.8678, 'grad_norm': 953.5204467773438, 'learning_rate': 5e-06, 'epoch': 0.01}                                                                                                                                                                                      
{'loss': 12.6281, 'grad_norm': 2949.16162109375, 'learning_rate': 5.833333333333334e-06, 'epoch': 0.01}                                                                                                                                                                       
{'loss': 11.5261, 'grad_norm': 2978.842529296875, 'learning_rate': 6.666666666666667e-06, 'epoch': 0.01}                                                                                                                                                                      
{'loss': 10.7645, 'grad_norm': 1403.374755859375, 'learning_rate': 7.500000000000001e-06, 'epoch': 0.01}                                                                                                                                                                      
{'loss': 17.2494, 'grad_norm': 2027.790283203125, 'learning_rate': 8.333333333333334e-06, 'epoch': 0.01}   
```

90-100 steps:
```
{'loss': 10.1056, 'grad_norm': 14.725829124450684, 'learning_rate': 1.961591468063112e-05, 'epoch': 0.12}                                                                                                                                                                     
{'loss': 5.5941, 'grad_norm': 9.888932228088379, 'learning_rate': 1.9604440636594863e-05, 'epoch': 0.12}                                                                                                                                                                      
{'loss': 5.6372, 'grad_norm': 27.609088897705078, 'learning_rate': 1.9592801175825435e-05, 'epoch': 0.12}                                                                                                                                                                     
{'loss': 8.1496, 'grad_norm': 13.072076797485352, 'learning_rate': 1.9580996498788602e-05, 'epoch': 0.12}                                                                                                                                                                     
{'loss': 5.4474, 'grad_norm': 10.157075881958008, 'learning_rate': 1.9569026808795647e-05, 'epoch': 0.12}                                                                                                                                                                     
{'loss': 5.5057, 'grad_norm': 16.48522186279297, 'learning_rate': 1.955689231199986e-05, 'epoch': 0.12}                                                                                                                                                                       
{'loss': 5.3535, 'grad_norm': 11.122950553894043, 'learning_rate': 1.954459321739298e-05, 'epoch': 0.12}                                                                                                                                                                      
{'loss': 5.4009, 'grad_norm': 16.74958038330078, 'learning_rate': 1.9532129736801616e-05, 'epoch': 0.13}                                                                                                                                                                      
{'loss': 5.2926, 'grad_norm': 12.336542129516602, 'learning_rate': 1.9519502084883585e-05, 'epoch': 0.13}                                                                                                                                                                     
{'loss': 5.3891, 'grad_norm': 25.544628143310547, 'learning_rate': 1.9506710479124212e-05, 'epoch': 0.13} 
```

Reply:
> Out of curiosity, did you run experiments where the base weights were not updated?

The final results comparison is shown in the table below. As we can observe, there is a significant drop in performance when using PiSSA without updating the base weights.
|  | loss | gsm8k | math |
| --- | --- | --- | --- |
| base model+PiSSA | 1.776114435263083 | 0.33965125094768767 | 0.0538 |
| residual model+PiSSA | 0.2408599260521912 | 0.7149355572403336 | 0.2416 |